allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+        while token_answer_span_end < len(tokenized_context) and (
+            tokenized_context[token_answer_span_end].idx is None
+            or (
+                tokenized_context[token_answer_span_end].idx
+                + len(sanitize_wordpiece(tokenized_context[token_answer_span_end].text))
+                <= answer_end
+            )
+        ):
+            token_answer_span_end += 1
+        # If token_answer_span_end points beyond the end of the context, this just chooses all tokens up to the
+        # last one.
+        assert token_answer_span_end > token_answer_span_start
+
+        # Tokenize the question
+        tokenized_question = self._tokenizer.tokenize(question)
+        tokenized_question = tokenized_question[: self.max_query_length]
+        tokenized_question = [
+            token._replace(type_id=self.non_content_type_id, idx=None)
+            for token in tokenized_question
+        ]
+
+        # Stride over the context, making instances
+        # Sequences are <s> question </s> </s> question </s>.
+        # This is technically not correct for anything but RoBERTa, but it does not affect the scores.
+        space_for_context = self.length_limit - len(tokenized_question) - 4
+        stride_start = 0
+        while True:
+            tokenized_context_window = tokenized_context[stride_start:]
+            tokenized_context_window = tokenized_context_window[:space_for_context]
+
+            window_token_answer_span = (
+                token_answer_span_start - stride_start,
+                token_answer_span_end - stride_start,
+            )
+            if any(i < 0 or i > len(tokenized_context_window) for i in window_token_answer_span):
+                # The answer is not contained in the window.
+                window_token_answer_span = None
+
+            if not self.skip_invalid_examples or window_token_answer_span is not None:
+                additional_metadata = {"id": qid}
+                instance = self.text_to_instance(
+                    question,
+                    tokenized_question,
+                    context,
+                    tokenized_context_window,
+                    answers,
+                    window_token_answer_span,
+                    additional_metadata,
+                )
+                yield instance
+
+            stride_start += space_for_context
+            if stride_start >= len(tokenized_context):
+                break
+            stride_start -= self.stride
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        question: str,
+        tokenized_question: List[Token],
+        context: str,
+        tokenized_context: List[Token],
+        answers: List[str],
+        token_answer_span: Optional[Tuple[int, int]],
+        additional_metadata: Dict[str, Any] = None,
+    ) -> Instance:
+        fields = {}
+
+        # make the question field
+        def cls_token():
```

Comment:
Do these need to be functions?  I guess it's safer this way, though I don't think anything would break if you used the same token multiple times, particularly because they are immutable objects.

Reply:
No, they do not. I tried it, and it looks safe without. ba458b00895f1e2f3b4f0af3e798aa80f05613ab